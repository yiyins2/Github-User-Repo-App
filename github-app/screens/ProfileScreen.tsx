import * as React from 'react';
import { StyleSheet } from 'react-native';
import {
  Avatar, Headline, Title, Button, List, Paragraph, ActivityIndicator,
} from 'react-native-paper';
import Icon from 'react-native-vector-icons/FontAwesome';
import { useState, useEffect } from 'react';
import { BottomTabNavigationProp } from '@react-navigation/bottom-tabs';
import { Route } from '@react-navigation/native';
import getUser from '../github-model/userQuery';
import User from '../github-model/GithubUser';
import { Text, View } from '../components/Themed';
import { BottomTabParamList, ProfileParamList } from '../types';

const styles = StyleSheet.create({
  container: {
    alignItems: 'center',
    justifyContent: 'center',
    marginTop: 35,
  },
  listItem: {
    marginLeft: 10,
    marginBottom: -15,
    marginTop: -15,
  },
  name: {
    marginTop: 10,
    paddingTop: 10,
    fontSize: 35,
  },
  bio: {
    padding: 10,
    paddingLeft: 20,
    paddingRight: 15,
  },
});

/**
 * @param param0 navigation function and route params
 * @returns the Profile screen
 */
export default function ProfileScreen({ navigation, route } :
{ navigation : BottomTabNavigationProp<BottomTabParamList, 'Profile'>,
  route : Route<'Profile', ProfileParamList['ProfileScreen']> }) : JSX.Element {
  const [user, setUser] = useState<null | User>(null);
  const [errorMsg, setErrorMsg] = useState<null | string>(null);

  /**
   * Initialize the Profile data
   */
  useEffect(() => {
    getUser(route.params?.login).then(
      (response) => { setUser(response); },
      (error : Error) => { setErrorMsg(error.message); },
    );
  }, [route.params?.login]);

  if (errorMsg !== null) {
    return <View style={styles.container}>{errorMsg}</View>;
  }

  // render the view
  return (
    <View>
      { user === null ? <View style={styles.container}><ActivityIndicator animating size="large" /></View>
        : (
          <View>
            <View style={styles.container}>
              <Avatar.Image size={160} source={{ uri: user.avatarUrl }} />
              <Headline style={styles.name}>{user.name}</Headline>
              <View style={{ flexDirection: 'row', alignItems: 'center', marginBottom: 8 }}>
                <Icon style={{ marginRight: 6, marginTop: 2 }} name="github" size={24} />
                <Title>{user.login}</Title>
              </View>
              <View style={{ flexDirection: 'row' }}>
                <Button
                  style={{ margin: 5 }}
                  mode="outlined"
                  onPress={() => navigation.navigate('Followers', { params: { login: user.login }, screen: 'FollowersScreen' })}
                >
                  {user.followersCount}
                  {' '}
                  followers
                </Button>
                <Button
                  style={{ margin: 5 }}
                  mode="outlined"
                  onPress={() => navigation.navigate('Following', { params: { login: user.login }, screen: 'FollowingScreen' })}
                >
                  {user.followingCount}
                  {' '}
                  following
                </Button>
              </View>
              <Button
                style={{ margin: 5 }}
                mode="outlined"
                onPress={() => navigation.navigate('Repositories', { params: { login: user.login }, screen: 'RepositoriesScreen' })}
              >
                {user.repositoriesCount}
                {' '}
                repositories
              </Button>
              <Paragraph style={styles.bio}>{user.bio}</Paragraph>
            </View>
            <List.Item style={styles.listItem} title={user.email} left={() => <List.Icon icon="email" />} />
            <List.Item style={styles.listItem} title={user.websiteUrl} left={() => <List.Icon icon="link-variant" />} />
            <Text style={styles.bio}>
              Profile created at
              {' '}
              {user.createdAt.toLocaleDateString()}
              .
            </Text>
          </View>
        )}
    </View>
  );
}
