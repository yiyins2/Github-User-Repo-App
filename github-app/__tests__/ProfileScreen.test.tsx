/* eslint-disable @typescript-eslint/require-await */
/* eslint-disable @typescript-eslint/no-explicit-any */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React from 'react';
import renderer, { act } from 'react-test-renderer';
import axios from 'axios';
// import { render, fireEvent } from '@testing-library/react-native/';
import ProfileScreen from '../screens/ProfileScreen';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;

test('renders normal response correctly', async () => {
  mockedAxios.post.mockResolvedValue({
    data: {
      data: {
        user: {
          avatarUrl: 'https://avatars.githubusercontent.com/u/31669157?s=200&u=6b0e5bb8e1425816f333d5889d0538300938d302&v=4',
          name: 'a',
          login: 'b',
          bio: 'c',
          websiteUrl: 'd',
          email: 'e',
          following: {
            totalCount: 1,
          },
          followers: {
            totalCount: 2,
          },
          createdAt: '9/5/2017',
          repositories: {
            totalCount: 3,
          },
        },
      },
    },
  });
  let tree : renderer.ReactTestRenderer | undefined;
  const navigation : any = { navigate: jest.fn() };
  const route: any = { params: { login: jest.fn() } };
  await act(async () => {
    tree = renderer.create(<ProfileScreen navigation={navigation} route={route} />);
  });
  expect(tree?.toJSON()).toMatchSnapshot();
});

test('renders undefined response with activity indicator', async () => {
  mockedAxios.post.mockResolvedValue(undefined);
  let tree : renderer.ReactTestRenderer | undefined;
  const navigation : any = { navigate: jest.fn() };
  const route: any = { params: { login: jest.fn() } };
  await act(async () => {
    tree = renderer.create(<ProfileScreen navigation={navigation} route={route} />);
  });
  expect(tree?.toJSON()).toMatchSnapshot();
});

test('renders undefined error response with error message', async () => {
  mockedAxios.post.mockRejectedValue(new Error('Some Error Occurred'));
  let tree : renderer.ReactTestRenderer | undefined;
  const navigation : any = { navigate: jest.fn() };
  const route: any = { params: { login: jest.fn() } };
  await act(async () => {
    tree = renderer.create(<ProfileScreen navigation={navigation} route={route} />);
  });
  expect(tree?.toJSON()).toMatchSnapshot();
});
