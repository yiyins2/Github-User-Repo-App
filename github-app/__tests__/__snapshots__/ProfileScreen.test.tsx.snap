// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders normal response correctly 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      undefined,
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#fff",
        },
        undefined,
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#fff",
          },
          Object {
            "alignItems": "center",
            "justifyContent": "center",
            "marginTop": 35,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#6200ee",
              "borderRadius": 80,
              "height": 160,
              "width": 160,
            },
            undefined,
          ]
        }
      >
        <Image
          source={
            Object {
              "uri": "https://avatars.githubusercontent.com/u/31669157?s=200&u=6b0e5bb8e1425816f333d5889d0538300938d302&v=4",
            }
          }
          style={
            Object {
              "borderRadius": 80,
              "height": 160,
              "width": 160,
            }
          }
        />
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 24,
                  "letterSpacing": 0,
                  "lineHeight": 32,
                  "marginVertical": 2,
                },
                Object {
                  "fontSize": 35,
                  "marginTop": 10,
                  "paddingTop": 10,
                },
              ],
            ],
          ]
        }
      >
        a
      </Text>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "marginBottom": 8,
            },
          ]
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": undefined,
                "fontSize": 24,
              },
              Object {
                "marginRight": 6,
                "marginTop": 2,
              },
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          ï‚›
        </Text>
        <Text
          style={
            Array [
              Object {
                "color": "#000000",
                "fontFamily": "System",
                "fontWeight": "400",
              },
              Object {
                "textAlign": "left",
              },
              Array [
                Object {
                  "textAlign": "left",
                },
                Object {
                  "color": "rgba(0, 0, 0, 0.87)",
                  "fontFamily": "System",
                  "fontWeight": "500",
                  "writingDirection": "ltr",
                },
                Array [
                  Object {
                    "fontSize": 20,
                    "letterSpacing": 0.15,
                    "lineHeight": 30,
                    "marginVertical": 2,
                  },
                  undefined,
                ],
              ],
            ]
          }
        >
          b
        </Text>
      </View>
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#fff",
            },
            Object {
              "flexDirection": "row",
            },
          ]
        }
      >
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "rgba(0, 0, 0, 0.29)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "elevation": 0,
              "margin": 5,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#6200ee",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                2
                 
                followers
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "backgroundColor": "transparent",
              "borderColor": "rgba(0, 0, 0, 0.29)",
              "borderRadius": 4,
              "borderStyle": "solid",
              "borderWidth": 0.5,
              "elevation": 0,
              "margin": 5,
              "minWidth": 64,
            }
          }
        >
          <View
            accessibilityRole="button"
            accessibilityState={
              Object {
                "disabled": undefined,
              }
            }
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Array [
                Object {
                  "overflow": "hidden",
                },
                Object {
                  "borderRadius": 4,
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "center",
                  },
                  undefined,
                ]
              }
            >
              <Text
                numberOfLines={1}
                selectable={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "400",
                    },
                    Object {
                      "textAlign": "left",
                    },
                    Array [
                      Object {
                        "letterSpacing": 1,
                        "marginHorizontal": 16,
                        "marginVertical": 9,
                        "textAlign": "center",
                      },
                      undefined,
                      Object {
                        "textTransform": "uppercase",
                      },
                      Object {
                        "color": "#6200ee",
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      Object {
                        "fontFamily": "System",
                        "fontWeight": "500",
                      },
                      undefined,
                    ],
                  ]
                }
              >
                1
                 
                following
              </Text>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "borderColor": "rgba(0, 0, 0, 0.29)",
            "borderRadius": 4,
            "borderStyle": "solid",
            "borderWidth": 0.5,
            "elevation": 0,
            "margin": 5,
            "minWidth": 64,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 4,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                undefined,
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#6200ee",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    undefined,
                  ],
                ]
              }
            >
              3
               
              repositories
            </Text>
          </View>
        </View>
      </View>
      <Text
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "textAlign": "left",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.87)",
                "fontFamily": "System",
                "fontWeight": "400",
                "writingDirection": "ltr",
              },
              Array [
                Object {
                  "fontSize": 14,
                  "letterSpacing": 0.25,
                  "lineHeight": 20,
                  "marginVertical": 2,
                },
                Object {
                  "padding": 10,
                  "paddingLeft": 20,
                  "paddingRight": 15,
                },
              ],
            ],
          ]
        }
      >
        c
      </Text>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          false,
          Array [
            Object {
              "padding": 8,
            },
            Object {
              "marginBottom": -15,
              "marginLeft": 10,
              "marginTop": -15,
            },
          ],
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "margin": 8,
                "width": 40,
              },
              undefined,
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "lineHeight": 24,
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                ],
                Object {
                  "fontFamily": "material-community",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ó°‡®
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 16,
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                  },
                  undefined,
                ],
              ]
            }
          >
            e
          </Text>
        </View>
      </View>
    </View>
    <View
      accessible={true}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          false,
          Array [
            Object {
              "padding": 8,
            },
            Object {
              "marginBottom": -15,
              "marginLeft": 10,
              "marginTop": -15,
            },
          ],
        ]
      }
    >
      <View
        style={
          Object {
            "flexDirection": "row",
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Array [
              Object {
                "alignItems": "center",
                "height": 40,
                "justifyContent": "center",
                "margin": 8,
                "width": 40,
              },
              undefined,
            ]
          }
        >
          <Text
            accessibilityElementsHidden={true}
            allowFontScaling={false}
            importantForAccessibility="no-hide-descendants"
            pointerEvents="none"
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontSize": 24,
                },
                Array [
                  Object {
                    "lineHeight": 24,
                    "transform": Array [
                      Object {
                        "scaleX": 1,
                      },
                    ],
                  },
                  Object {
                    "backgroundColor": "transparent",
                  },
                ],
                Object {
                  "fontFamily": "material-community",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            ó°Œ¹
          </Text>
        </View>
        <View
          style={
            Array [
              Object {
                "marginVertical": 6,
                "paddingLeft": 8,
              },
              Object {
                "flex": 1,
                "justifyContent": "center",
              },
            ]
          }
        >
          <Text
            numberOfLines={1}
            selectable={false}
            style={
              Array [
                Object {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontWeight": "400",
                },
                Object {
                  "textAlign": "left",
                },
                Array [
                  Object {
                    "fontSize": 16,
                  },
                  Object {
                    "color": "rgba(0, 0, 0, 0.87)",
                  },
                  undefined,
                ],
              ]
            }
          >
            d
          </Text>
        </View>
      </View>
    </View>
    <Text
      style={
        Array [
          Object {
            "color": "#000",
          },
          Object {
            "padding": 10,
            "paddingLeft": 20,
            "paddingRight": 15,
          },
        ]
      }
    >
      Profile created at
       
      9/5/2017
      .
    </Text>
  </View>
</View>
`;

exports[`renders undefined error response with error message 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginTop": 35,
      },
    ]
  }
>
  Some Error Occurred
</View>
`;

exports[`renders undefined response with activity indicator 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "#fff",
      },
      Object {
        "alignItems": "center",
        "justifyContent": "center",
        "marginTop": 35,
      },
    ]
  }
>
  Cannot read property 'data' of undefined
</View>
`;
